darwin/torch/dataset.py

```diff
@@ -172,79 +104,86 @@ def _map_annotation(self, index: int):
-        self.subsample(percentage)
+    def __getitem__(self, index: int):
+        # Load images and masks
+        img = load_pil_image(self.images_path[index])
+        target = self._map_annotation(index)
+        if self.transforms is not None:
+            img, target = self.transforms(img, target)
+        return img, target
 
     def __len__(self):
         return len(self.images_path)
 
     def __str__(self):
-        format_string = (
-            f"{self.__class__.__name__}():\n"
-            f"  Root: {self.root}\n"
-            f"  Number of images: {len(self.images_path)}"
-        )
-        return format_string
+        return (f"{self.__class__.__name__}():\n"
+                f"  Root: {self.dataset.local_path}\n"
+                f"  Number of images: {len(self.images_path)}")
 
 
 class ClassificationDataset(Dataset):
-    def __init__(self, root: Path, image_set: str, split_id: Optional[str] = None, transforms=None):
-        super(ClassificationDataset, self).__init__(root, image_set, split_id, transforms)
-        self.classes = [e.strip() for e in open(str(root / "lists/classes_tags.txt"))]
+    def __init__(self, dataset, split: Path, transforms: Optional[List] = None):
+        """See superclass for documentation"""
+        super().__init__(dataset=dataset, split=split, transforms=transforms)
+        self.classes = [e.strip() for e in open(str(self.dataset.local_path / "lists/classes_tags.txt"))]
 
     def _map_annotation(self, index: int):
+        """See superclass for documentation"""
         with open(str(self.annotations_path[index])) as f:
             annotation = json.load(f)["annotations"]
-            for obj in annotation:
-                if "tag" in obj:
-                    return {"category_id": self.classes.index(obj["name"])}
+        return {"category_id": [self.classes.index(a["name"]) for a in annotation if "tag" in a]}
 
 
 class InstanceSegmentationDataset(Dataset):
-    def __init__(self, root: Path, image_set: str, split_id: Optional[str] = None, transforms=None):
-        super(InstanceSegmentationDataset, self).__init__(root, image_set, split_id, transforms)
-        self.classes = [e.strip() for e in open(str(root / "lists/classes_masks.txt"))]
+    def __init__(self, dataset, split: Path, transforms: Optional[List] = None):
+        """See superclass for documentation"""
+        super().__init__(dataset=dataset, split=split, transforms=transforms)
+        self.classes = [e.strip() for e in open(str(self.dataset.local_path / "lists/classes_masks.txt"))]
+        # Prepend the default transform to convert polygons to instance masks
+        if self.transforms is None:
+            self.transforms = []
+        self.transforms.insert(0, T.ConvertPolysToInstanceMasks())
+        self.transforms = T.Compose(transforms)
 
     def _map_annotation(self, index: int):
+        """See superclass for documentation"""
         with open(str(self.annotations_path[index])) as f:
```

Comment:
try this instead: `with self.annotation_path[index].open() as f`

Reply:
Done